### ISSUE TITLE
Add delta calculation method

### ISSUE THREAD

**Issue by Crazz-Zaac** [Crazz-Zaac/informative-missingness#2]
- Add a method to properly handle *delta* calculation to the data pre-processing pipeline
- The data should be horizontally stacked before feeding to the model
-

**Comment by Kcm5750** [comment]
class DeltaFeatureProcessor:
    """
    A class to handle delta calculations and feature combinations for time series data
    """
    
    def __init__(self, 
                 x_columns: List[str],
                 m_columns: List[str],
                 id_column: str = 'id',
                 time_column: str = 'time',
                 delta_periods: int = 1):
        """
        Initialize the DeltaFeatureProcessor
        
        Parameters:
        x_columns: List of column names for X features
        m_columns: List of column names for M features  
        id_column: Column name for identifier
        time_column: Column name for time index
        delta_periods: Number of periods for delta calculation
        """
        self.x_columns = x_columns
        self.m_columns = m_columns
        self.id_column = id_column
        self.time_column = time_column
        self.delta_periods = delta_periods
        self.scaler = StandardScaler()
        self.feature_combinations = {}
        
    def calculate_deltas(self, df: pd.DataFrame, columns: List[str]) -> pd.DataFrame:
        """
        Calculate delta (difference) values for specified columns
        
        Parameters:
        df: Input DataFrame
        columns: Columns to calculate deltas for
        
        Returns:
        DataFrame with delta columns added
        """
        df = df.copy()
        
        # Ensure proper sorting by time within each group
        if self.id_column in df.columns and self.time_column in df.columns:
            df = df.sort_values([self.id_column, self.time_column])
        
        for col in columns:
            if col in df.columns:
                # Calculate delta within each group if id_column exists
                if self.id_column in df.columns:
                    df[f'delta_{col}'] = df.groupby(self.id_column)[col].diff(self.delta_periods)
                else:
                    df[f'delta_{col}'] = df[col].diff(self.delta_periods)
        
        return df
    
    def create_feature_combinations(self, df: pd.DataFrame) -> Dict[str, pd.DataFrame]:
        """
        Create all specified feature combinations
        
        Parameters:
        df: Input DataFrame with original and delta features
        
        Returns:
        Dictionary of DataFrames with different feature combinations
        """
        # Get all available columns
        all_columns = df.columns.tolist()
        
        # Filter out non-feature columns
        non_feature_cols = [self.id_column, self.time_column] if self.id_column in df.columns else []
        if self.time_column in df.columns:
            non_feature_cols.append(self.time_column)
        
        feature_columns = [col for col in all_columns if col not in non_feature_cols]
        
        # Create feature combinations
        combinations = {}
        
        # 1. Original features only (x, m)
        x_features = [col for col in self.x_columns if col in feature_columns]
        m_features = [col for col in self.m_columns if col in feature_columns]
        
        combinations['x'] = df[non_feature_cols + x_features].copy()
        combinations['m'] = df[non_feature_cols + m_features].copy()
        combinations['x_m'] = df[non_feature_cols + x_features + m_features].copy()
        
        # 2. Delta features only
        delta_x_features = [f'delta_{col}' for col in self.x_columns if f'delta_{col}' in feature_columns]
        delta_m_features = [f'delta_{col}' for col in self.m_columns if f'delta_{col}' in feature_columns]
        
        combinations['delta'] = df[non_feature_cols + delta_x_features + delta_m_features].copy()
        combinations['delta_x'] = df[non_feature_cols + delta_x_features].copy()
        combinations['delta_m'] = df[non_feature_cols + delta_m_features].copy()
        
        # 3. Combined features (original + delta)
        combinations['x_delta'] = df[non_feature_cols + x_features + delta_x_features].copy()
        combinations['m_delta'] = df[non_feature_cols + m_features + delta_m_features].copy()
        combinations['x_m_delta'] = df[non_feature_cols + x_features + m_features + delta_x_features + delta_m_features].copy()
        
        # 4. All features
        combinations['all'] = df.copy()
        
        self.feature_combinations = combinations
        return combinations
    
    def horizontal_stack_features(self, df: pd.DataFrame, combination_type: str = 'x_m_delta') -> np.ndarray:
        """
        Horizontally stack features for model input
        
        Parameters:
        df: Input DataFrame
        combination_type: Type of feature combination to use
        
        Returns:
        Numpy array with horizontally stacked features
        """
        if combination_type not in self.feature_combinations:
            self.create_feature_combinations(df)
        
        combination_df = self.feature_combinations[combination_type]
        
        # Remove non-feature columns for stacking
        non_feature_cols = []
        if self.id_column in combination_df.columns:
            non_feature_cols.append(self.id_column)
        if self.time_column in combination_df.columns:
            non_feature_cols.append(self.time_column)
        
        feature_df = combination_df.drop(columns=non_feature_cols, errors='ignore')
        
        # Convert to numpy array and ensure proper shape
        stacked_features = feature_df.values
        
        return stacked_features
    
    def preprocess_pipeline(self, 
                          df: pd.DataFrame,
                          target_column: Optional[str] = None,
                          scale_features: bool = True) -> Dict:
        """
        Complete preprocessing pipeline with delta calculation and feature combinations
        
        Parameters:
        df: Input DataFrame
        target_column: Optional target column for supervised learning
        scale_features: Whether to scale features
        
        Returns:
        Dictionary containing processed DataFrames and feature arrays
        """
        # Step 1: Calculate deltas for both X and M features
        print("Calculating delta features...")
        df_with_deltas = self.calculate_deltas(df, self.x_columns + self.m_columns)
        
        # Step 2: Handle missing values created by diff operation
        df_with_deltas = df_with_deltas.dropna()
        
        # Step 3: Scale features if requested
        if scale_features:
            print("Scaling features...")
            feature_columns = [col for col in df_with_deltas.columns 
                             if col not in [self.id_column, self.time_column, target_column] 
                             and col is not None]
            
            if feature_columns:
                df_with_deltas[feature_columns] = self.scaler.fit_transform(df_with_deltas[feature_columns])
        
        # Step 4: Create all feature combinations
        print("Creating feature combinations...")
        feature_combinations = self.create_feature_combinations(df_with_deltas)
        
        # Step 5: Prepare results
        results = {
            'processed_df': df_with_deltas,
            'feature_combinations': feature_combinations,
            'stacked_features': {}
        }
        
        # Step 6: Create horizontally stacked arrays for each combination
        for combo_name in feature_combinations.keys():
            results['stacked_features'][combo_name] = self.horizontal_stack_features(
                df_with_deltas, combo_name
            )
        
        return results

# Example usage and test function
def test_delta_processor():
    """
    Test function to demonstrate the DeltaFeatureProcessor
    """
    # Create sample data
    np.random.seed(42)
    n_samples = 100
    n_ids = 5
    
    data = {
        'id': np.repeat(range(1, n_ids + 1), n_samples // n_ids),
        'time': np.tile(range(n_samples // n_ids), n_ids),
        'x1': np.random.randn(n_samples),
        'x2': np.random.randn(n_samples),
        'm1': np.random.randn(n_samples),
        'm2': np.random.randn(n_samples),
        'target': np.random.randn(n_samples)
    }
    
    df = pd.DataFrame(data)
    
    # Initialize processor
    processor = DeltaFeatureProcessor(
        x_columns=['x1', 'x2'],
        m_columns=['m1', 'm2'],
        id_column='id',
        time_column='time',
        delta_periods=1
    )
    
    # Run preprocessing pipeline
    results = processor.preprocess_pipeline(df, target_column='target', scale_features=True)
    
    # Display results
    print("Original DataFrame shape:", df.shape)
    print("Processed DataFrame shape:", results['processed_df'].shape)
    print("\nAvailable feature combinations:")
    for combo_name, combo_df in results['feature_combinations'].items():
        print(f"{combo_name}: {combo_df.shape}")
    
    print("\nStacked feature shapes:")
    for combo_name, stacked_array in results['stacked_features'].items():
        print(f"{combo_name}: {stacked_array.shape}")
    
    return results

# Integration with your existing pipeline
def enhanced_processing_pipeline(input_file: str, 
                               output_dir: str,
                               x_columns: List[str],
                               m_columns: List[str],
                               id_column: str = 'id',
                               time_column: str = 'time') -> Dict:
    """
    Enhanced processing pipeline with delta feature handling
    """
    import os
    os.makedirs(output_dir, exist_ok=True)
    
    # Load your data (adjust based on your data format)
    if input_file.endswith('.csv'):
        df = pd.read_csv(input_file)
    elif input_file.endswith('.jsonl'):
        df = pd.read_json(input_file, lines=True)
    else:
        raise ValueError("Unsupported file format")
    
    # Initialize delta processor
    processor = DeltaFeatureProcessor(
        x_columns=x_columns,
        m_columns=m_columns,
        id_column=id_column,
        time_column=time_column
    )
    
    # Process data with delta features
    results = processor.preprocess_pipeline(df, scale_features=True)
    
    # Save results
    results['processed_df'].to_csv(os.path.join(output_dir, 'processed_with_deltas.csv'), index=False)
    
    # Save each feature combination
    for combo_name, combo_df in results['feature_combinations'].items():
        combo_df.to_csv(os.path.join(output_dir, f'features_{combo_name}.csv'), index=False)
    
    # Save stacked features as numpy arrays
    stacked_dir = os.path.join(output_dir, 'stacked_features')
    os.makedirs(stacked_dir, exist_ok=True)
    
    for combo_name, stacked_array in results['stacked_features'].items():
        np.save(os.path.join(stacked_dir, f'stacked_{combo_name}.npy'), stacked_array)
    
    print("Processing complete! All feature combinations saved.")
    return results

# Example configuration for your specific use case
if __name__ == "__main__":
    # Test the processor
    test_results = test_delta_processor()
    
    # Example of how to use with your actual data
    # Replace with your actual column names
    x_columns = ['feature1', 'feature2', 'feature3']  # Your X features
    m_columns = ['metric1', 'metric2']
